File: src/Bindings/Logging.cpp
#include "pluginsdk/bridgemain.h"
#include "pluginsdk/_plugins.h"
#include "Marshal.hpp"

namespace Dotx64Dbg::Native
{
    public ref class Logging
    {
    public:
        static void Write(System::String^ text)
        {
            auto textStr = interop::toUTF8(text);
            _plugin_logprint(textStr.c_str());
        }

        static void WriteLine(System::String^ text)
        {
            auto textStr = interop::toUTF8(text);
            textStr = textStr + "\n";
            _plugin_logprint(textStr.c_str());
        }

        static void WriteLine()
        {
            _plugin_logprint("\n");
        }
    };
}

File: src/Dotx64Managed/Logging.cs
using System;
using System.IO;
using System.Text;

namespace Dotx64Dbg
{
    class ConsoleWriter : TextWriter
    {
        public override Encoding Encoding { get { return Encoding.UTF8; } }

        public override void Write(string value)
        {
            Native.Logging.Write(value);
        }

        public override void WriteLine(string value)
        {
            Native.Logging.WriteLine(value);
        }
        public override void WriteLine()
        {
            Native.Logging.WriteLine();
        }
    }

    public static class Logging
    {
        internal static void Initialize()
        {
            var sink = new ConsoleWriter();
            Console.SetOut(sink);
            Console.SetError(sink);
        }
    }

}
